### PR TITLE
Improve step to catch SOAP exceptions and add functional tests for it

### PR DESCRIPTION
**Pull request by BR0kEN-**
#7

### CODE REVIEW

File: src/Context/SoapContext.php [hunk 1]
```diff
@@ -102,6 +108,22 @@ public function iRegisterXpathNamespaces(TableNode $namespaces)
     }
 
     /**
+     * @Given /^(?:|I )expect SOAP exception(?:| with code "(\d+)")(?:|( and| or)? with message "([^"]+?)")$/
```
Comment (asgorobets) [inline src/Context/SoapContext.php]: Will it make more sense to use @Then keyword for this step? It sounds more like an assertion step then a setup step. I know behat ignores it, but for the user ;)

Comment (BR0kEN-) [inline src/Context/SoapContext.php]: "I love you" :D

File: src/Context/SoapContext.php [hunk 2]
```diff
@@ -102,6 +108,22 @@ public function iRegisterXpathNamespaces(TableNode $namespaces)
...
+    public function expectException($code = null, $condition = null, $message = null)
+    {
+        // Exit with an error because we're expected an exception and got nothing.
+        if (null === $this->fault) {
+            throw new \RuntimeException('Expected \SoapFault exception was not thrown!');
+        }
+
+        new SoapFaultProcessor($this->fault, $code, $message, $condition);
+
+        // If processor was not thrown an exception then we shouldn't too.
```
Comment (asgorobets) [inline src/Context/SoapContext.php]: WIll it sound better if we reword it to `If processor didn't throw an exception, then we shouldn't too`?

File: src/Context/SoapContext.php [hunk 3]
```diff
@@ -215,4 +237,18 @@ public function savedValueNotMatchesRegExp($pattern)
     {
         Assertions::assertNotRegExp($pattern, $this->value);
     }
+
+    /**
+     * @BeforeStep
+     */
+    public function beforeStep(BeforeStepScope $scope)
```
Comment (asgorobets) [inline src/Context/SoapContext.php]: Can we rename the method to explicitly say that we're working with exceptions? Like `beforeStepCheckException`?

File: src/Utils/SoapFaultProcessor.php [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+     * Process expected error message.
+     *
+     * @return $this
+     */
+    private function processMessage()
+    {
+        if (null !== $this->message) {
+            $value = $this->exception->getMessage();
+
+            if (strpos($value, trim($this->message)) === false) {
```
Comment (asgorobets) [inline src/Utils/SoapFaultProcessor.php:106]: Awesome, but I see cases when we might need a regexp pattern to match the error message, we could even make the step behave both for 'contains' and 'matches' by just  testing the message string for start and end of regular expression delimiter `/` for example. However this can be added later in next PR ;)

File: src/Utils/SoapFaultProcessor.php [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+     * Filter messages depending on condition.
+     */
+    private function processCondition()
+    {
+        if (!empty($this->condition)) {
+            $exceptions = count($this->messages);
+
+            // - At least one message needed to be able to choose one from and meet "or" condition.
+            // - Two messages are needed to be able to meet "and" condition.
+            if (('or' === $this->condition && $exceptions < 2) || ('and' === $this->condition && $exceptions < 1)) {
```
Comment (asgorobets) [inline src/Utils/SoapFaultProcessor.php]: ('and' === $this->condition && $exceptions === 0i) may work as well as  `< 1`, but in my opinion is more clear. We're saying the 'and' means no exceptions, both conditions should pass

File: src/Utils/SoapFaultProcessor.php [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    /**
+     * Filter messages depending on condition.
+     */
+    private function processCondition()
+    {
+        if (!empty($this->condition)) {
+            $exceptions = count($this->messages);
+
+            // - At least one message needed to be able to choose one from and meet "or" condition.
+            // - Two messages are needed to be able to meet "and" condition.
```
Comment (asgorobets) [inline src/Utils/SoapFaultProcessor.php]: Maybe we say `No failed assertions should be present in order to meet 'and' condition`?

File: src/Utils/SoapFaultProcessor.php (hunk 5 above)
Comment (BR0kEN-) [inline src/Utils/SoapFaultProcessor.php]: Disagree. `< 1` is pretty easy for understanding.

File: src/Utils/SoapFaultProcessor.php (hunk 4 above)
Comment (BR0kEN-) [inline src/Utils/SoapFaultProcessor.php:106]: This needed to be discussed. I cannot imagine a situation when we need a regex and definitely sure that allow to input regexes will bring butthurt for us.

**Comment by asgorobets** [comment]
PHP 5.4 is failing but It's ok as we dropped it in master, merging.
Thanks for your contribution!
